Pytorch training updated + LRP

faroukmokhtar opened the pull request.

jpata commented on mlpf/pytorch_delphes/LRP/hooks.py:
```diff
@@ -0,0 +1,106 @@
+use_gpu = torch.cuda.device_count()>0
+multi_gpu = torch.cuda.device_count()>1
+
+try:
+    if not ("CUDA_VISIBLE_DEVICES" in os.environ):
+        import setGPU
+        if multi_gpu:
+            print('Will use multi_gpu..')
+            print("Let's use", torch.cuda.device_count(), "GPUs!")
+        else:
+            print('Will use single_gpu..')
+except Exception as e:
+    print("Could not import setGPU, running CPU-only")
+
+#define the global base device
+if use_gpu:
+    device = torch.device('cuda:0')
+else:
+    device = torch.device('cpu')
```
this seems to be repeated in several places, can you run this GPU detection once from a common file?

jpata commented on mlpf/pytorch_delphes/LRP/hooks.py:
```diff
@@ -0,0 +1,106 @@
+from torch.utils.data import random_split
+import torch_cluster
+
+sys.path.insert(1, '../')
+sys.path.insert(1, '../../../plotting/')
+sys.path.insert(1, '../../../mlpf/plotting/')
```
I think this pattern could give some issues with python modules. You should have proper modules defined by placiong `__init__.py` files in the appropriate locations, and be able to do relative imports as in e.g. https://github.com/jpata/particleflow/blob/master/mlpf/tfmodel/model.py#L8

jpata commented on mlpf/pytorch_delphes/LRP/hooks.py:
```diff
@@ -0,0 +1,106 @@
+    return self.fc1(self.flatten(x))
+
+
+net = myNet()
```
is this a test file (in which case perhaps the file should be moved/removed, and the following code put in `__main__`), or a library file to be imported elsewhere (in which case, this code should be removed)?

jpata commented on mlpf/pytorch_delphes/graph_data_delphes.py:
```diff
@@ -7,15 +7,17 @@
 from torch_geometric.data import Dataset, Data, Batch
 import itertools
 from glob import glob
+import numba
```
is numba actually used?

jpata commented on mlpf/pytorch_delphes/gravnet.py:
```diff
@@ -0,0 +1,122 @@
+except ImportError:
+    knn = None
+
+# copied it from pytorch_geometric source code
+# ADDED: retrieve edge_index, retrieve edge_weight
+# CHANGED: self.lin -> self.lin_p
+# CHANGED: used reduce='sum' instead of reduce='mean' in the message passing
+# REMOVED: skip connection
+
+class GravNetConv(MessagePassing):
```
this seems duplicated with respect to gravnet_LRP.py, can you make a common import?

jpata commented on scripts/local_test_delphes_pytorch.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+# !/bin/bash
```
typo?

jpata commented on scripts/local_test_delphes_pytorch.sh:
```diff
@@ -10,32 +10,44 @@ mkdir -p data/pythia8_ttbar
-# predict on some test data and make plots
-python3 ../mlpf/pytorch/eval_end2end_delphes.py --dataset data/pythia8_ttbar \
-  --path experiments/PFNet* --model PFNet7 --start 1 --stop 2 --epoch 1 --target gen
+echo Begining the training..
```
```suggestion
echo Beginning the training..
```

jpata reviewed (commented):
Thanks for the PR! Since we are now a bit more people, let me take the occasion to make some detailed comments to improve the state of the codebase and to exercise good software engineering practices.

1. Could you please provide a PR description with some details what you are changing and implementing here?
1. Could you please comment on the choice of making the pytorch model now specific for the delphes dataset? Can this be made generic, so we could use other datasets as well, as for the TF model?
1. It seems there are a few repeated patterns that could be improved in the code
   1. GPU detection and device choice in a common file, to be imported elsewhere (in the beginning of the main pipeline entry point ideally)
   2. reduce the amount of duplicate definitions for the models (e.g. the GravNet layer)
   2. remove `sys.path.insert` workaround, add `__init__.py` files to make sure module imports can be done cleanly
   3. removing executable code (or putting it in `if__name__ == "__main__"`) from files that are imported elsewhere (see point 4)
   4. commenting code in/out to enable inference could perhaps be improved via a flag (see point 4)
1. Could a common entry point (e.g. similar to https://github.com/jpata/particleflow/blob/master/mlpf/pipeline.py) be defined, from which the various training, inference and validation codes can be run?
2. Could you please make sure the basic pytorch functionality (dataset generation, training, inference) is being exercised in the github CI pipeline, and the pipeline runs successfully?
3. Can you check if `numba` is actually used in the pytorch code, and if it's not used, remove the import (it is a somewhat heavy dependency)?

faroukmokhtar commented:
Thanks for your suggestions! I removed many unnecessary files to avoid confusion. Here are some answers to the points above:

1) PR description: added a pipeline.py file to train on delphes dataset, make predictions and make evaluation plots. Furthermore, added an LRP machinery that computes relevance scores to explain the model predictions. Check the README.md for more information.

2) Currently the model works perfectly with delphes. It should be straightforward to test if it works with cms, and make simple changes it doesn't). I will work on this when I get back.

3) Worked on most of these. Still have to: 
- add __init__.py as a workaround sys.path.insert
- make gpu detection check in only one file (the main file to run). Some files still need to get updated in this regard. Specifically, the LRP files.

4) Running pipeline.py (the rename of training.py) with different flags will either train a model, make predictions on data, and/or make evaluation plots.

5) All checks have passed now.

6) Removed numba import.

Let me know if you have any more suggestions.

jpata commented:
Thanks, this is much improved. If the LRP does not work in the test environment:

```
 ./scripts/local_test_delphes_pytorch.sh: line 61:  2244 Killed                  python3 lrp_pipeline.py --n_test=1 --batch_size=4 --lrp_dataset_qcd='../test_tmp_delphes/data/pythia8_qcd' --lrp_outpath='../test_tmp_delphes/experiments/' --lrp_load_model='PFNet7_gen_ntrain_1_nepochs_10_batch_size_4_lr_0.0001_alpha_0.0002_both__nn1_nn3'
Error: Process completed with exit code 137.
```

could you remove it from `scripts/local_test_delphes_pytorch.sh`?

faroukmokhtar commented:
Done. Now it passed all the github checks.

Let me know if there's anything else to edit/improve to merge the PR.

jpata commented:
Thanks! Merging this now.
